File: packages/ld-adapter/src/write/semantic-patch.ts
import {
	SemanticPatchInstructionSchema,
	type RolloutIntent,
	type SemanticPatchInstruction,
	type TargetingIntent,
} from '@ff-promo/contracts';

export type ResolvedRolloutIds = {
	treatmentVariationId: string;
	controlVariationId: string;
	ruleId?: string;
};

export function buildRolloutWeights(
	treatmentThousandths: number,
	treatmentVariationId: string,
	controlVariationId: string,
): Record<string, number> {
	if (treatmentThousandths < 0 || treatmentThousandths > 100_000) {
		throw new Error('treatmentThousandths must be between 0 and 100000');
	}
	const controlThousandths = 100_000 - treatmentThousandths;
	const weights = {
		[treatmentVariationId]: treatmentThousandths,
		[controlVariationId]: controlThousandths,
	};
	const sum = Object.values(weights).reduce((acc, value) => acc + value, 0);
	if (sum !== 100_000) {
		throw new Error(`rollout weights must sum to 100000, got ${sum}`);
	}
	return weights;
}

function rolloutInstruction(
	rollout: RolloutIntent,
	resolved: ResolvedRolloutIds,
): SemanticPatchInstruction {
	const rolloutWeights = buildRolloutWeights(
		rollout.treatmentPercentThousandths,
		resolved.treatmentVariationId,
		resolved.controlVariationId,
	);

	if (rollout.mode === 'rule') {
		if (!resolved.ruleId) {
			throw new Error('ruleId required when rollout.mode is rule');
		}
		return SemanticPatchInstructionSchema.parse({
			kind: 'updateRuleVariationOrRollout',
			ruleId: resolved.ruleId,
			rolloutWeights,
			rolloutBucketBy: rollout.rolloutBucketBy,
			rolloutContextKind: rollout.rolloutContextKind,
		});
	}

	return SemanticPatchInstructionSchema.parse({
		kind: 'updateFallthroughVariationOrRollout',
		rolloutWeights,
		rolloutBucketBy: rollout.rolloutBucketBy,
		rolloutContextKind: rollout.rolloutContextKind,
	});
}

export function buildTargetingPatchBody(
	intent: TargetingIntent,
	resolved?: ResolvedRolloutIds,
): {
	environmentKey: string;
	comment?: string;
	instructions: Array<Record<string, unknown>>;
} {
	const instructions: Array<Record<string, unknown>> = [];

	if (intent.turnOn === false) {
		instructions.push({ kind: 'turnFlagOff' });
	} else if (intent.turnOn ?? true) {
		instructions.push({ kind: 'turnFlagOn' });
	}

	if (intent.rollout) {
		if (!resolved) {
			throw new Error('resolved variation IDs required when rollout is present');
		}
		const instruction = rolloutInstruction(intent.rollout, resolved);
		const { kind, ...rest } = instruction;
		instructions.push({ kind, ...rest });
	}

	for (const instruction of instructions) {
		if (instruction.kind === 'updatePercentageRollout') {
			throw new Error('updatePercentageRollout is not a valid LD instruction');
		}
	}

	return {
		environmentKey: intent.environmentKey,
		comment: intent.comment,
		instructions,
	};
}

File: kotlin/modules/ld-adapter/src/main/kotlin/com/ffpromo/ldadapter/write/SemanticPatch.kt
package com.ffpromo.ldadapter.write

import com.ffpromo.contracts.RolloutIntent
import com.ffpromo.contracts.SemanticPatchInstruction
import com.ffpromo.contracts.TargetingIntent
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

data class ResolvedRolloutIds(
    val treatmentVariationId: String,
    val controlVariationId: String,
    val ruleId: String? = null,
)

fun buildRolloutWeights(
    treatmentThousandths: Int,
    treatmentVariationId: String,
    controlVariationId: String,
): Map<String, Int> {
    require(treatmentThousandths in 0..100_000) {
        "treatmentThousandths must be between 0 and 100000"
    }
    val controlThousandths = 100_000 - treatmentThousandths
    val weights = mapOf(
        treatmentVariationId to treatmentThousandths,
        controlVariationId to controlThousandths,
    )
    val sum = weights.values.sum()
    check(sum == 100_000) { "rollout weights must sum to 100000, got $sum" }
    return weights
}

private fun rolloutInstruction(
    rollout: RolloutIntent,
    resolved: ResolvedRolloutIds,
): SemanticPatchInstruction {
    val rolloutWeights = buildRolloutWeights(
        rollout.treatmentPercentThousandths,
        resolved.treatmentVariationId,
        resolved.controlVariationId,
    )

    return if (rollout.mode == "rule") {
        val ruleId = resolved.ruleId ?: error("ruleId required when rollout.mode is rule")
        SemanticPatchInstruction.UpdateRuleVariationOrRollout(
            ruleId = ruleId,
            rolloutWeights = rolloutWeights,
            rolloutBucketBy = rollout.rolloutBucketBy,
            rolloutContextKind = rollout.rolloutContextKind,
        )
    } else {
        SemanticPatchInstruction.UpdateFallthroughVariationOrRollout(
            rolloutWeights = rolloutWeights,
            rolloutBucketBy = rollout.rolloutBucketBy,
            rolloutContextKind = rollout.rolloutContextKind,
        )
    }
}

fun buildTargetingPatchBody(
    intent: TargetingIntent,
    resolved: ResolvedRolloutIds? = null,
): JsonObject {
    val instructions = buildJsonArray {
        when (intent.turnOn) {
            false -> add(buildJsonObject { put("kind", JsonPrimitive("turnFlagOff")) })
            null, true -> add(buildJsonObject { put("kind", JsonPrimitive("turnFlagOn")) })
        }

        val rollout = intent.rollout
        if (rollout != null) {
            requireNotNull(resolved) { "resolved variation IDs required when rollout is present" }
            add(instructionToJson(rolloutInstruction(rollout, resolved)))
        }
    }

    instructions.forEach { element ->
        val kind = element.jsonObject["kind"]?.jsonPrimitive?.content
        if (kind == "updatePercentageRollout") {
            error("updatePercentageRollout is not a valid LD instruction")
        }
    }

    return buildJsonObject {
        put("environmentKey", JsonPrimitive(intent.environmentKey))
        intent.comment?.let { put("comment", JsonPrimitive(it)) }
        put("instructions", instructions)
    }
}

private fun instructionToJson(instruction: SemanticPatchInstruction): JsonObject =
    when (instruction) {
        SemanticPatchInstruction.TurnFlagOn -> buildJsonObject { put("kind", JsonPrimitive("turnFlagOn")) }
        SemanticPatchInstruction.TurnFlagOff -> buildJsonObject { put("kind", JsonPrimitive("turnFlagOff")) }
        is SemanticPatchInstruction.UpdateFallthroughVariationOrRollout -> buildJsonObject {
            put("kind", JsonPrimitive("updateFallthroughVariationOrRollout"))
            put("rolloutWeights", weightsToJson(instruction.rolloutWeights))
            put("rolloutBucketBy", JsonPrimitive(instruction.rolloutBucketBy))
            put("rolloutContextKind", JsonPrimitive(instruction.rolloutContextKind))
        }
        is SemanticPatchInstruction.UpdateRuleVariationOrRollout -> buildJsonObject {
            put("kind", JsonPrimitive("updateRuleVariationOrRollout"))
            put("ruleId", JsonPrimitive(instruction.ruleId))
            put("rolloutWeights", weightsToJson(instruction.rolloutWeights))
            put("rolloutBucketBy", JsonPrimitive(instruction.rolloutBucketBy))
            put("rolloutContextKind", JsonPrimitive(instruction.rolloutContextKind))
        }
    }

private fun weightsToJson(weights: Map<String, Int>) = buildJsonObject {
    weights.forEach { (key, value) -> put(key, JsonPrimitive(value)) }
}
